app/src/sharedTest/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivityTest.kt

```diff
@@ -320,10 +341,335 @@ class QuestionPlayerActivityTest {
+      openHintsAndSolutionDialog()
+      clickRevealNewHintAndSol(hintAndSolIndex = 1, isSolution = true)
+
+      onView(withText(context.getString(R.string.this_will_reveal_the_solution))).inRoot(isDialog())
+        .check(
+          matches(isDisplayed())
+        )
+    }
+  }
+
+  @Test
+  fun testQuestionPlayer_revealSolution_configChange_dialogBoxVisible() {
+    launchForSkillList(SKILL_ID_LIST_1).use {
+      makeFirstNewHintsVisible(SKILL_ID_1)
+
+      openHintsAndSolutionDialog()
+      clickRevealNewHintAndSol(hintAndSolIndex = 0, isSolution = false)
+      navigateBackToQuestionPlayer()
+
+      makeSecondNewHintAndSolutionVisible()
+      openHintsAndSolutionDialog()
+      clickRevealNewHintAndSol(hintAndSolIndex = 1, isSolution = true)
+
+      rotateToLandscape()
+
+      onView(withText(context.getString(R.string.this_will_reveal_the_solution))).inRoot(isDialog())
+        .check(
+          matches(isDisplayed())
+        )
+    }
+  }
+
+  @Test
+  fun testQuestionPlayer_revealSolution_dialogBoxVisible_clickReveal_solutionVisible() {
+    launchForSkillList(SKILL_ID_LIST_1).use {
+      makeFirstNewHintsVisible(SKILL_ID_1)
+
+      openHintsAndSolutionDialog()
+      clickRevealNewHintAndSol(hintAndSolIndex = 0, isSolution = false)
+      navigateBackToQuestionPlayer()
+
+      makeSecondNewHintAndSolutionVisible()
+      openHintsAndSolutionDialog()
+      clickRevealNewHintAndSol(hintAndSolIndex = 1, isSolution = true)
+
+      onView(withText(context.getString(R.string.reveal))).inRoot(isDialog())
+        .perform(click())
+
+      testCoroutineDispatchers.runCurrent()
+      onView(isRoot()).check(
+        matches(
+          not(
+            ViewMatchers.withSubstring("The only solution is")
+          )
+        )
+      )
+    }
+  }
+
+  @Test
+  fun testQuestionPlayer_revealSolution_dialogBoxVisible_clickCancel_solutionNotVisible() {
+    launchForSkillList(SKILL_ID_LIST_1).use {
+      makeFirstNewHintsVisible(SKILL_ID_1)
+
+      openHintsAndSolutionDialog()
+      clickRevealNewHintAndSol(hintAndSolIndex = 0, isSolution = false)
+      navigateBackToQuestionPlayer()
+
+      makeSecondNewHintAndSolutionVisible()
+      openHintsAndSolutionDialog()
+      clickRevealNewHintAndSol(hintAndSolIndex = 1, isSolution = true)
+
+      onView(withText(context.getString(R.string.cellular_data_alert_dialog_cancel_button)))
+        .inRoot(isDialog())
+        .perform(click())
+
+      testCoroutineDispatchers.runCurrent()
+      onView(withText("Reveal Solution")).inRoot(isDialog()).check(matches(isDisplayed()))
+    }
+  }
+
   private fun setUpTestApplicationComponent() {
     ApplicationProvider.getApplicationContext<TestApplication>().inject(this)
   }
 
+  /**
+   * Makes a new hint visible by submitting two wrong answers
+   * provided hint is available and no answer is submitted until
+   * this function is completely executes.
+   */
+  private fun makeFirstNewHintsVisible(skillId: String) {
```

Comment:
is this applicable for all skill id? or does it work only for specific skill id?

Reply:
This should work for all skill ids if it is called within 60 sec and no answer is submitted to the question till that point.